src/nbl/ext/DepthPyramidGenerator/DepthPyramidGenerator.cpp

```diff
@@ -0,0 +1,319 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+// TODO: make it work for multiple passes
+
+#include <nbl/ext/DepthPyramidGenerator/DepthPyramidGenerator.h>
+
+using namespace nbl;
+using namespace nbl::core;
+using namespace nbl::asset;
+using namespace nbl::video;
+
+namespace nbl
+{
+namespace ext
+{
+namespace DepthPyramidGenerator
+{
+
+DepthPyramidGenerator::DepthPyramidGenerator(IVideoDriver* driver, IAssetManager* am, core::smart_refctd_ptr<IGPUImageView> inputDepthImageView,
+		const Config& config)
+	: m_driver(driver), m_config(config)
+{
+	const char* source =
+		R"(#version 430 core
+#define WORKGROUP_X_AND_Y_SIZE %u
+#define MIPMAP_LEVELS_PER_PASS %u
+#define MIP_IMAGE_FORMAT %s
+#define STRETCH_MIN
+#define %s
+
+layout(local_size_x = WORKGROUP_X_AND_Y_SIZE, local_size_y = WORKGROUP_X_AND_Y_SIZE) in;
+ 
+#include "../../../include/nbl/builtin/glsl/ext/DepthPyramidGenerator/depth_pyramid_generator_impl.glsl"
+)";
+
+	constexpr char* imageFormats[] =
+	{
+		"r16f", "r32f", "rg16f", "rg32f"
+	};
+
+	const char* format;
+	switch (config.outputFormat)
+	{
+	case E_IMAGE_FORMAT::EIF_R16_FLOAT:
+		format = imageFormats[0];
+		break;
+	case E_IMAGE_FORMAT::EIF_R32_FLOAT:
+		format = imageFormats[1];
+		break;
+	case E_IMAGE_FORMAT::EIF_R16G16_FLOAT:
+		format = imageFormats[2];
+		break;
+	case E_IMAGE_FORMAT::EIF_R32G32_FLOAT:
+		format = imageFormats[3];
+		break;
+	default:
+		assert(false);
+	}
+
+	constexpr char* redOps[] =
+	{
+		"REDUCION_OP_MIN", "REDUCION_OP_MAX", "REDUCION_OP_BOTH"
+	};
+
+	const char* redOp;
+	switch (config.op)
+	{
+	case E_MIPMAP_GENERATION_OPERATOR::EMGO_MIN:
+		redOp = redOps[0];
+		break;
+	case E_MIPMAP_GENERATION_OPERATOR::EMGO_MAX:
+		redOp = redOps[1];
+		break;
+	case E_MIPMAP_GENERATION_OPERATOR::EMGO_BOTH:
+		redOp = redOps[2];
+		break;
+	default:
+		assert(false);
+	}
+
+	const uint32_t perPassMipCnt = static_cast<uint32_t>(config.workGroupSize) == 32u ? 6u : 5u;
+
+	constexpr size_t extraSize = 32u;
+	auto shaderCode = core::make_smart_refctd_ptr<ICPUBuffer>(strlen(source) + extraSize + 1u);
+	snprintf(reinterpret_cast<char*>(shaderCode->getPointer()), shaderCode->getSize(), source, static_cast<uint32_t>(m_config.workGroupSize), perPassMipCnt, format, redOp);
+
+	auto cpuSpecializedShader = core::make_smart_refctd_ptr<asset::ICPUSpecializedShader>(
+		core::make_smart_refctd_ptr<asset::ICPUShader>(std::move(shaderCode), asset::ICPUShader::buffer_contains_glsl),
+		asset::ISpecializedShader::SInfo{ nullptr, nullptr, "main", asset::ISpecializedShader::ESS_COMPUTE });
+
+	auto gpuShader = driver->createGPUShader(core::smart_refctd_ptr<const asset::ICPUShader>(cpuSpecializedShader->getUnspecialized()));
```

Comment:
I believe @sadiuk made a nice function for creating shaders with format strings (Cause we seem to be doing it quite often)

Reply:
yeah but it's in the tessendorf waves pr

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/169/commits/aff989018702873b7865244310b6022cd76a719e#diff-16bd39c1e8e1502d45bbc8d218649271623ba6e92a790b3e8f1f9c21a1a2aeffR86